### PR TITLE
For CFFI, remove variadic arguments

### LINKED ISSUE #30986: chip-repl CFFI Remove variadic arguments from python to C
In `src/controller/python/chip/clusters/Command.py` and `src/controller/python/chip/clusters/Attribute.py` we rely on variadic arguments to pass along command/writeattribut arguments.

There has to be a better way to do this and we should look into this eventually

### PR DESCRIPTION
**Pull request by tehampson**
Fixes: https://github.com/project-chip/connectedhomeip/issues/30986

Removing variadic arguments will make creating `pychip_CommandSender_TestOnlySendBatchCommands` that will share common implementation with `pychip_CommandSender_SendBatchCommands` easier to implement and maintain.

Additionally this seems like a more clean way to pass these arguments to c++

Arguable we could follow precedent with passing structures where python side uses `construct` and we have a packed structure on the c++ side. But I recently discovered that we can share a common struct over CFFI using `ctypes.Structure`. This seems easier and is a more sane way to build and pass the structure through.

Eventually we should go back and fix all structures to use this, but for now I want this PR to focus on removal of variadic arguments when going from Python to C

### CODE REVIEW

**Review by tcarmelveilleux — APPROVED** [review]

File: src/controller/python/chip/clusters/attribute.cpp [hunk 1]
```diff
@@ -256,13 +257,12 @@ struct __attribute__((packed)) PyReadAttributeParams
     bool autoResubscribe;
 };
 
-// Encodes n attribute write requests, follows 3 * n arguments, in the (AttributeWritePath*=void *, uint8_t*, size_t) order.
 PyChipError pychip_WriteClient_WriteAttributes(void * appContext, DeviceProxy * device, size_t timedWriteTimeoutMsSizeT,
-                                               size_t interactionTimeoutMsSizeT, size_t busyWaitMsSizeT, size_t n, ...);
+                                               size_t interactionTimeoutMsSizeT, size_t busyWaitMsSizeT,
+                                               chip::python::PyWriteAttributeData * writeAttributesData, size_t n);
```
Comment (tehampson) [inline src/controller/python/chip/clusters/attribute.cpp:262]: Rename all the `size_t n` to something a little more clear

File: src/controller/python/chip/interaction_model/Delegate.h [hunk 2]
```diff
@@ -23,8 +23,27 @@
...
 namespace chip {
 namespace python {
+
 static constexpr ClusterStatus kUndefinedClusterStatus = 0xFF;
-}
+
+// This needs to match the python definition that uses the same name.
+struct PyInvokeRequestData
+{
+    void * mCommandPath;
```
Comment (andy31415) [inline src/controller/python/chip/interaction_model/Delegate.h:32]: should we specify the type of data this points to? void* is very generic. Same question for mAttributePath below.

File: src/controller/python/chip/clusters/attribute.cpp [hunk 3]
```diff
@@ -352,35 +355,30 @@ PyChipError pychip_WriteClient_WriteAttributes(void * appContext, DeviceProxy *
...
 
+    for (size_t i = 0; i < attributeDataLength; i++)
     {
-        for (size_t i = 0; i < n; i++)
+        void * path   = writeAttributesData[i].mAttributePath;
+        void * tlv    = writeAttributesData[i].mTlvData;
+        size_t length = writeAttributesData[i].mTlvLength;
+
+        python::AttributePath pathObj;
+        memcpy(&pathObj, path, sizeof(python::AttributePath));
```
Comment (andy31415) [inline src/controller/python/chip/clusters/attribute.cpp:367]: I find it somewhat confusing that we use python::AttributePath which is different than python::Controller::AttributePath (the latter having version) from data build from a sconstruct that is AttributePathIBStruct.

How can we guarantee this is ok ? IBStruct has data version however python::AttributePath has it. We probably do not care, however binary compatibility enforcement here is a bit painful. Should we comment on this and why everything is ok?

Comment (andy31415) [reply src/controller/python/chip/clusters/attribute.cpp:367]: Was looking at the wrong part ... Controller one has no version.

Overall maybe we should add comments in the sconstruct definitions that these MUST match binary wise. I even wonder if there is a way to test and enforce this.

**Review by andy31415 — APPROVED** [review]

File: src/controller/python/chip/interaction_model/Delegate.h (hunk 2 above)
Comment (tehampson) [reply src/controller/python/chip/interaction_model/Delegate.h]: It is no longer a pointer and just a struct

**Pull request merged**